### PR TITLE
Extend statistics

### PR DESCRIPTION
**Pull request by FrNecas**
This PR adds the following statistics:
 - number of instructions compared
 - number of instructions where 1:1 comparison was used (i.e. semantic preserving pattern was not used)
 - number of C code lines analyzed

Tracking the lines analyzed requires a set and the calculations on top of this data-structure resulted in a 10 second (260 -> 270) increase when analyzing the RHEL KABI. This feels quite significant, therefore it has to be explicitly enabled using `--extended-stat` option. The other statistics are just simple counters, therefore, they are included in the base statistics.

Example output comparing RHEL8 kernels:
```
Lines compared:          36537
Instructions compared:   109760
1:1 equal instructions:  107380 (98%)
```

### CODE REVIEW

File: diffkemp/semdiff/caching.py [hunk 1]
```diff
@@ -55,13 +55,17 @@ class Vertex:
...
+                     inst_cnt=None, inst_equal_cnt=None, lines_cnt=None):
             self.names = names
             self.result = result
             self.nonfun_diffs = []
             self.successors = ([], [])
             self.files = files
             self.lines = lines
+            self.inst_cnt = inst_cnt
+            self.inst_equal_cnt = inst_equal_cnt
+            self.lines_cnt = lines_cnt
```
Comment (viktormalik) [inline diffkemp/semdiff/caching.py:68]: I'm wondering if it's worth enclosing these into a separate object like `stats`. It could save some lines, especially if we add more stats in future, and also better represents the purpose of these fields.

The encapsulation would probably have to be done in SimpLL, too, so it's a bit of extra work.

Comment (FrNecas) [reply diffkemp/semdiff/caching.py:68]: Good idea, I will change this. I am not really sure about saving some lines. The `Stats` object/class in the frontend could load the sub-yaml dynamically but I think it will still be a bit more maintainable to explicitly state the `Stats` attributes in the frontend.

Comment (FrNecas) [reply diffkemp/semdiff/caching.py]: Done

**Review by viktormalik — COMMENTED** [review]
LGTM, thanks!

One thing I'm thinking about is to move instructions counts behind `--extended-stats`, too. Not that it'd hard to compute, it's just that the numbers are intended for our experiments and don't seem much useful for regular users. In fact, the same applies for detailed statistics about compared objects (see `Result.report_object_stat`). Maybe we could move those under `--extended-stats`, too, as a part of this PR?

Also, could you please squash the first and the third commit together and add some better commit messages?

Thanks!

File: diffkemp/cli.py [hunk 2]
```diff
@@ -105,6 +105,10 @@ def make_argument_parser():
     compare_ap.add_argument("--report-stat",
                             help="report statistics of the analysis",
                             action="store_true")
+    compare_ap.add_argument("--extended-stat",
+                            help="compute extended statistics that are more "
+                                 "expensive to compute (e.g. line count). ",
```
Comment (viktormalik) [inline diffkemp/cli.py:110]: Maybe rephrase this?
```suggestion
                            help="report detailed statistics about the analysis "
                                 "(may be more expensive to compute). ",
```

File: diffkemp/simpll/ModuleComparator.cpp [hunk 3]
```diff
@@ -104,6 +104,18 @@ void ModuleComparator::compareFunctions(Function *FirstFun,
...
+    ComparedFuns.at({FirstFun, SecondFun}).First.stats.instEqualCnt =
+            fComp.InstEqual;
+    ComparedFuns.at({FirstFun, SecondFun}).First.stats.linesCnt =
+            fComp.ComparedLinesL.size();
+    ComparedFuns.at({FirstFun, SecondFun}).Second.stats.instCnt =
+            fComp.ComparedInstR;
+    ComparedFuns.at({FirstFun, SecondFun}).Second.stats.instEqualCnt =
+            fComp.InstEqual;
+    ComparedFuns.at({FirstFun, SecondFun}).Second.stats.linesCnt =
+            fComp.ComparedLinesR.size();
```
Comment (viktormalik) [inline diffkemp/simpll/ModuleComparator.cpp:118]: Could we enclose this into a method/function? Also repeats below.

**Review by viktormalik — COMMENTED** [review]
I'm getting division by zero:
```
$ bin/diffkemp compare snapshots/linux-4.18.0-80.el8 snapshots/linux-4.18.0-147.el8 -f __alloc_pages_nodemask --stdout --report-stat --extended-stat
[...]
Traceback (most recent call last):
  File "/home/vmalik/src/diffkemp/diffkemp/bin/diffkemp", line 6, in <module>
    diffkemp.run_from_cli()
  File "/home/vmalik/.local/lib/python3.11/site-packages/diffkemp/diffkemp.py", line 37, in run_from_cli
    args.func(args)
  File "/home/vmalik/.local/lib/python3.11/site-packages/diffkemp/diffkemp.py", line 503, in compare
    result.report_stat(args.show_errors, args.extended_stat)
  File "/home/vmalik/.local/lib/python3.11/site-packages/diffkemp/semdiff/result.py", line 292, in report_stat
    self.report_object_stat()
  File "/home/vmalik/.local/lib/python3.11/site-packages/diffkemp/semdiff/result.py", line 250, in report_object_stat
    equal_percent = equal_instructions / compared_instructions * 100
                    ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
ZeroDivisionError: division by zero
```

**Comment by FrNecas** [comment]
I am not really sure what to output if there were no instructions compared. I've used `0 (nan%)` for now. Let me know if you think of something better. Or maybe we could leave out this statistic completely if the number of instructions is 0? I slightly lean towards being explicit and consistent between all runs (i.e. always output all statistics).

File: diffkemp/semdiff/result.py [hunk 4]
```diff
@@ -259,6 +269,10 @@ def __hash__(self):
             print("Elapsed time:            {:.2f} s".format(
                 self.stop_time - self.start_time))
         print("Functions compared:      {}".format(compared))
+        print("Lines compared:          {}".format(compared_lines))
+        print("Instructions compared:   {}".format(compared_instructions))
+        print("1:1 equal instructions:  {0} ({1:.0f}%)".format(
+            equal_instructions, equal_percent))
```
Comment (viktormalik) [inline diffkemp/semdiff/result.py:275]: Let's insert one blank line here (before `Total differences:`) to visually separate the following block.

**Review by viktormalik — COMMENTED** [review]
> I am not really sure what to output if there were no instructions compared. I've used 0 (nan%) for now. Let me know if you think of something better. Or maybe we could leave out this statistic completely if the number of instructions is 0? I slightly lean towards being explicit and consistent between all runs (i.e. always output all statistics).

Agreed, let's be explicit. My example works now, I'm just wondering why the bug occurred when there's:
```
Lines compared:          0
Instructions compared:   10570
1:1 equal instructions:  10390 (98%)
```
Was there a bug before? Also I'm curious why the number of lines compared is 0.

Anyways, the update LGTM, just one small nit for better readability of the extended stats.

**Comment by FrNecas** [comment]
> Was there a bug before? Also I'm curious why the number of lines compared is 0.

Are you sure that you are running the same example? I am still getting 0 compared instructions and 0 analyzed lines. Weird...

**Comment by viktormalik** [comment]
> > Was there a bug before? Also I'm curious why the number of lines compared is 0.
> 
> Are you sure that you are running the same example? I am still getting 0 compared instructions and 0 analyzed lines. Weird...

Hm, I'm getting 0 everywhere now, too, but I'm sure that I ran the same example (just checked my terminal history). Strange... Anyways, there are 466 functions compared and no instructions so something is clearly off here.

**Review by viktormalik — CHANGES_REQUESTED** [review]

**Review by viktormalik — APPROVED** [review]
Wonderful!
